xml/value.go

```diff
@@ -0,0 +1,329 @@
+package xml
+
+import (
+	"encoding/base64"
+	"fmt"
+	"math"
+	"math/big"
+	"strconv"
+)
+
+// Value represents an XML Value type
+// XML Value types: Object, Array, Map, String, Number, Boolean.
+type Value struct {
+	w       writer
+	scratch *[]byte
+
+	// xml start element is the associated start element for the Value
+	startElement StartElement
+
+	// indicates if the Value represents a flattened shape
+	isFlattened bool
+}
+
+// newValue returns a new Value encoder. newValue does NOT write the start element tag
+func newValue(w writer, scratch *[]byte, startElement StartElement) Value {
+	return Value{
+		w:            w,
+		scratch:      scratch,
+		startElement: startElement,
+	}
+}
+
+// newWrappedValue writes the start element xml tag and returns a Value
+func newWrappedValue(w writer, scratch *[]byte, startElement StartElement) Value {
+	writeStartElement(w, startElement)
+	return Value{w: w, scratch: scratch, startElement: startElement}
+}
+
+// writeStartElement takes in a start element and writes it.
+// It handles namespace, attributes in start element.
+func writeStartElement(w writer, el StartElement) error {
+	if el.isZero() {
+		return fmt.Errorf("xml start element cannot be nil")
+	}
```

Comment:
This returns an error but will never be caught.  Maybe the encoder types should have a once set error value that will allow errors to be bubbled up?  Right now this error will be silently ignored.

Reply:
As discussed online, punting this task for now. Will add a TODO instead. Need to investigate a better way to bubble up the error and not require multiple error checks at each level